src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -183,8 +192,34 @@ fn compile<'a, 'cfg: 'a>(cx: &mut Context<'a, 'cfg>,
     let (dirty, fresh, freshness) = if unit.profile.run_custom_build {
         custom_build::prepare(cx, unit)?
     } else {
-        let (freshness, dirty, fresh) = fingerprint::prepare_target(cx,
-                                                                         unit)?;
+        let mut work_override = None;
+        if unit.profile.check {
+            // If we are running `cargo check`, then we only need metadata from
+            // from the dependencies. The cheapest way to get that is by just
+            // building a metadata crate. However, if we have previously done
+            // a full build, then we can get the metadata from the rlib without
+            // doing any work. Therefore, we check here to see if such a lib
+            // exists and is fresh.
+            let mut lib_unit = unit.clone();
+            lib_unit.profile = cx.lib_profile(lib_unit.pkg.package_id());
+            let (freshness, lib_dirty, lib_fresh) = fingerprint::prepare_target(cx, &lib_unit)?;
+            if freshness == Freshness::Fresh {
```

Comment:
So historically the `fingerprint` module only checked whether one target was up to date, not checking dependencies. If, however, a dependency is recompiled we'll end up wanting to recompile this target even though the dependency information says it's fresh (as a dep changed).

I forget if `fingerprint` has since changed to take into account dependencies automatically, or whether that's still a concern of the `DependencyQueue`. Could you verify this?

In any case, the expected behavior for me would be that if you're running `cargo check` and there's an "fresh rlib" but it ends up needing recompiling anyway due to some other dep, we'd want to generate the rmeta instead of the rlib here.